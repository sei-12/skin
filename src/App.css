body {
	margin: 0;
	background-color: var(--base-bg);
    font-family: Hack Nerd Font Bold;
}

.tag-suggestion-window {
	background-color: var(--color7);
	width: 30vw;
    position: absolute;
    overflow: hidden;
}


.tag-suggestion-window-list-outer {
    position: relative;
    max-height: 15vh;
    overflow: scroll;
    border: solid 1px var(--color8);
}

.tag-suggestion-window-list-outer::-webkit-scrollbar {
	display: none;
}

.tag-suggestion-window-list-outer span {
    font-size: small
}

.tag-suggestion-unmatch {
    color: var(--color5);    
}

.tag-suggestion-match {
    color: var(--color6);    
}

.tag-suggestion-list-item {
    box-sizing: border-box;
    padding: 1px;
}

.focused-tag-suggestion-list-item {
    background-color: var(--color9);
}

:root {
    --color1: #455765;
    --color2: #c1e4e9;
    --color3: #393f4c;
    --color4: #20242f;
    --color5: rgb(150,150,150);
    --color6: white;
    --color7: #151a26; 
    --color8: rgb(20,20,20);
    --color9: rgb(31,36,48);
    --create-new-tag-window-bg: var(--color3);    
    --create-new-tag-bg: rgba(10,10,10,0.5);
    --base-bg: var(--color4);
    --defalut-input-box: var(--color2);
}