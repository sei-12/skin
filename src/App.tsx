import { Box, Button, Grid2, Input } from "@mui/material";
import { useState } from "react";
import { IData } from "./data";
import { BookmarkList } from "./components/BookmarkList";

function App() {

    const [items, setItems] = useState<IData.Bookmark[]>([
        {
            title: "hello1",
            key: crypto.randomUUID(),
            tags: ["tag1", "tag2", "tag3"],
            desc: "this is description".repeat(100),
            url: "decoy://aaa.www.ccc"
        },
        {
            title: "hello1".repeat(100),
            key: crypto.randomUUID(),
            tags: ["tag1", "tag2", "tag3"],
            desc: "this is description",
            url: "decoy://aaa.www.ccc"
        },
        {
            title: "hello1",
            key: crypto.randomUUID(),
            tags: ["tag1", "tag2", "tag3"],
            desc: "this is description",
            url: "decoy://aaa.www.ccc"
        },
        {
            title: "hello1",
            key: crypto.randomUUID(),
            tags: ["tag1", "tag2", "tag3"],
            desc: "this is description",
            url: "decoy://aaa.www.ccc"
        },
        {
            title: "hello1",
            key: crypto.randomUUID(),
            tags: ["tag1", "tag2", "tag3"],
            desc: "this is description",
            url: "decoy://aaa.www.ccc"
        },
        {
            title: "hello1",
            key: crypto.randomUUID(),
            tags: ["tag1", "tag2", "tag3"],
            desc: "this is description",
            url: "decoy://aaa.www.ccc"
        },
        {
            title: "hello3",
            key: crypto.randomUUID(),
            tags: ["tag5", "tag4", "tag3"],
            desc: "this is description".repeat(19),
            url: "decoy://aaa.www.ccc"
        },
        {
            title: "hello1",
            key: crypto.randomUUID(),
            tags: ["tag1", "tag2", "tag3"],
            desc: "this is description",
            url: "decoy://aaa.www.ccc"
        },
        {
            title: "hello1aaaa",
            key: crypto.randomUUID(),
            tags: ["tag1", "tag2", "tag3"],
            desc: "this is description".repeat(3),
            url: "decoy://aaa.www.ccc"
        },
        {
            title: "hello4",
            key: crypto.randomUUID(),
            tags: ["tag1", "tag2", "tag3"],
            desc: "this is description",
            url: "decoy://aaa.www.ccc"
        },
        {
            title: "hello11",
            key: crypto.randomUUID(),
            tags: ["tag1", "tag2", "tag3"],
            desc: "this is description",
            url: "decoy://aaa.www.ccc"
        },
        {
            title: "hello1",
            key: crypto.randomUUID(),
            tags: ["tag1", "tag2", "tag3"],
            desc: "this is description".repeat(8),
            url: "decoy://aaa.www.ccc"
        },
        {
            title: "hello1",
            key: crypto.randomUUID(),
            tags: ["tag1", "tag2", "tag3"],
            desc: "this is description",
            url: "decoy://aaa.www.ccc"
        },
        {
            title: "hello7",
            key: crypto.randomUUID(),
            tags: ["tag1", "tag2", "tag3"],
            desc: "this is description",
            url: "decoy://aaa.www.ccc"
        },
        {
            title: "hello1",
            key: crypto.randomUUID(),
            tags: ["tag1", "tag2", "tag3"],
            desc: "this is description",
            url: "decoy://aaa.www.ccc"
        },
        {
            title: "hello3",
            key: crypto.randomUUID(),
            tags: ["tag1", "tag2", "tag3"],
            desc: "this is description",
            url: "decoy://aaa.www.ccc"
        },
        {
            title: "hello3",
            key: crypto.randomUUID(),
            tags: ["tag1", "tag2", "tag3"],
            desc: "this is description",
            url: "decoy://aaa.www.ccc"
        },
    ])

    const [focusIndex, setFocusIndex] = useState(0)

    return (
        <Box
            sx={{ 
                width: "100vw", 
                height: "100vh",
                top:0,
                left: 0,
                overflow: "hidden",
                padding: 3,
                flexGrow: 1,
            }}
        >
            <Grid2
                container
                spacing={1}
                flexDirection={"column"}
                height={1}
                width={1}
            >
                <Grid2 size="auto">
                    <Input></Input>
                    <Button onClick={() => setFocusIndex(focusIndex + 1)}>down</Button>
                    <Button onClick={() => setFocusIndex(focusIndex - 1)}>up</Button>
                </Grid2> 

                <Grid2 
                    size="grow" 
                    sx={{ 
                        overflow: "hidden",
                        display: "flex",
                    }}
                >
                    <BookmarkList
                        focusIndex={focusIndex}
                        onClickEdit={() => {}}
                        onClickRemove={() => {}}
                        items={items}
                    ></BookmarkList>
                </Grid2>
            </Grid2>
        </Box>

    );
}

export default App;
