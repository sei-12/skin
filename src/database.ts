import { IData } from "./data";

export class DataBase implements IDataBase {
	
	
}

/**
 * テストなどで使う
 */
export interface IDataBase {
	insertBookmark(data: IData.Bookmark): Promise<void>
	deleteBookmark(id: number): Promise<void>
	insertTag(tag:string):Promise<void>

	isExistsTag(tag:string): Promise<boolean>
	findBookmark(tags: string[]): Promise<IData.Bookmark>
	findTag(predicate: string): Promise<string[]>
}
