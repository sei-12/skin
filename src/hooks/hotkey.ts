import { useHotkeysContext } from "react-hotkeys-hook";

//
// TODO
// ページを分けたから必要ない
// default,suggestion_windowの二つでいいと思う
//
export const HOTKEY_SCOPES = {
    SEARCH_BOOKMARK: "search_bookmark",
    SEARCH_BOOKMARK_SUGGESTION_WINDOW: "SEARCH_BOOKMARK_SUGGESTION_WINDOW",

    CREATE_NEW_BOOKMARK: "CREATE_NEW_BOOKMARK",
    CREATE_NEW_BOOKMARK_SUGGESTION_WINDOW: "CREATE_NEW_BOOKMARK_SUGGESTION_WINDOW",

    EDIT_BOOKMARK: "EDIT_BOOKMARK",
} as const;

type HOTKEY_SCOPES = (typeof HOTKEY_SCOPES)[keyof typeof HOTKEY_SCOPES]


export function useAppHotkey() {
    const hotkeysContext = useHotkeysContext()

    const switchScope = (to: HOTKEY_SCOPES) => {
        hotkeysContext.enabledScopes.forEach(scope => {
            hotkeysContext.disableScope(scope)
        })
        hotkeysContext.enableScope(to)
    }

    return {
        switchScope
    }
}