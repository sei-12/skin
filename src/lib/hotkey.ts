import { useHotkeysContext } from "react-hotkeys-hook";

export const HOTKEY_SCOPES = {
    SEARCH_BOOKMARK: "search_bookmark",
    SEARCH_BOOKMARK_SUGGESTION_WINDOW: "SEARCH_BOOKMARK_SUGGESTION_WINDOW",

    CREATE_NEW_BOOKMARK: "CREATE_NEW_BOOKMARK",
    CREATE_NEW_BOOKMARK_SUGGESTION_WINDOW: "CREATE_NEW_BOOKMARK_SUGGESTION_WINDOW",

    EDIT_BOOKMARK: "EDIT_BOOKMARK",
} as const;

type HOTKEY_SCOPES = (typeof HOTKEY_SCOPES)[keyof typeof HOTKEY_SCOPES]


export function useAppHotkey() {
    const hotkeysContext = useHotkeysContext()

    const switchScope = (to: HOTKEY_SCOPES) => {
        hotkeysContext.enabledScopes.forEach(scope => {
            hotkeysContext.disableScope(scope)
        })
        hotkeysContext.enableScope(to)
    }

    return {
        switchScope
    }
}