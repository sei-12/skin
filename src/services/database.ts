// import Database from "@tauri-apps/plugin-sql";
import { invoke } from "@tauri-apps/api/core";
import type { Bookmark } from "../../src-tauri/bindings/export/DbModels";

// TODO:エラーを通知する

export namespace DB {
    export async function insertBookmark(
        title: string,
        url: string,
        desc: string,
        tags: string[]
    ): Promise<void> {
        return await invoke<void>("insert_bookmark", { req: { title, url, desc, tags } })
            .catch((e) => { throw e })
    }

    export async function editBookmark(
        id: number,
        title: string,
        url: string,
        desc: string,
        tags: string[]
    ): Promise<void> {
        return await invoke("edit_bookmark", { req: { id, title, url, desc, tags } })
    }

    export async function getBookmark(id: number): Promise<Bookmark> {
        return await invoke<Bookmark>("get_bookmark", { id })
    }

    export async function findTag(predicate: string): Promise<string[]> {
        return await invoke<string[]>("find_tag", { predicate })
            .catch((e) => { throw e })
    }

    export async function deleteBookmark(id: number): Promise<void> {
        return await invoke<void>("delete_bookmark", { bookmarkId: id })
            .catch((e) => { throw e })
    }

    export async function fetchBookmarks(maxLength: number): Promise<Bookmark[]> {
        return await invoke<Bookmark[]>("fetch_bookmarks", { maxLength })
            .catch((e) => { throw e })
    }

    export async function isExistsTag(tag: string) {
        return await invoke<boolean>("is_exists_tag", { tag: tag })
            .catch((e) => { throw e })
    }

    export async function findBookmark(filterTags: string[]): Promise<Bookmark[]> {
        return await invoke<Bookmark[]>("find_bookmark", { filterTags })
            .catch((e) => { throw e })
    }

    export async function fuzzyFindTag(predicate: string) {
        return await invoke<[string, boolean][][]>("fuzzy_find_tag", { predicate })
    }
}