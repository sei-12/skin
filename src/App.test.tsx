import { act, fireEvent, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom/vitest";
import { describe, expect, test, vi } from "vitest";
import App from "./App";
import { HotkeysProvider } from "react-hotkeys-hook";
import { HOTKEY_SCOPES } from "./hotkey";
import userEvent from "@testing-library/user-event";
import { IDataBase } from "./database";
import { IData } from "./data";

const dbConnection:IDataBase = vi.hoisted(() => ({
    async insertBookmark(_title: string, _url: string, _desc: string , _tags: string[]): Promise<void> { },
    async deleteBookmark(_id: number): Promise<void> { },
    async insertTag(_tag: string): Promise<void> { },
    async isExistsTag(_tag: string): Promise<boolean> {
        return true
    },
    async findBookmark(_tags: string[]): Promise<IData.Bookmark[]> {
        return []
    },
    async findTag(_predicate: string): Promise<string[]> {
        return [
            "a",
            "b",
            "c"
        ]
    }
}))

vi.mock("./database",() => ({
    dbConnection
}))

describe("App",() => {
    test("hotkey",async () => {
        const {asFragment} = render(
            <HotkeysProvider initiallyActiveScopes={[HOTKEY_SCOPES.SEARCH_BOOKMARK]}>
                <App></App>
            </HotkeysProvider>
        )
        
        await act(async () => {
            fireEvent.keyUp(document, { key: '/' });
        })
        
        let inputPredicateBox = screen.getByPlaceholderText("/")
        expect(inputPredicateBox).toHaveFocus()
        
        
        
        // MEMO: このテストは動かなくなる予定。（タグ検索のロジックがでこいだから）
        await act(async () => {
            userEvent.type(inputPredicateBox,"a")
        })
        
        expect(asFragment()).toMatchSnapshot()
        
    })
})