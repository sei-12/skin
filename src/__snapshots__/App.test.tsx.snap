// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > hotkey 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-kvkzl0"
  >
    <div
      class="MuiGrid2-root MuiGrid2-container MuiGrid2-direction-xs-row MuiGrid2-spacing-xs-1 css-l8gw4m-MuiGrid2-root"
    >
      <div
        class="MuiGrid2-root MuiGrid2-direction-xs-row MuiGrid2-grid-xs-auto css-1djy14n-MuiGrid2-root"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1588512-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
            data-testid="AddBoxIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2m-2 10h-4v4h-2v-4H7v-2h4V7h2v4h4z"
            />
          </svg>
        </button>
        <div
          class="MuiBox-root css-1sgu0ab"
        >
          <div
            class="MuiBox-root css-1qqo5vm"
          >
            <div
              class="MuiFormControl-root MuiTextField-root css-x5lmpp-MuiFormControl-root-MuiTextField-root"
            >
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-focused MuiInputBase-formControl css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  autocomplete="off"
                  class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
                  id=":r0:"
                  placeholder="/"
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-w4cd9x"
                  >
                    <span
                      aria-hidden="true"
                      class="notranslate"
                    >
                      â€‹
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <div
              class="MuiBox-root css-971pk6"
            />
          </div>
        </div>
      </div>
      <div
        class="MuiGrid2-root MuiGrid2-direction-xs-row MuiGrid2-grid-xs-grow css-36r1yx-MuiGrid2-root"
      >
        <div
          class="MuiStack-root css-7jr0kh-MuiStack-root"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
